### PR TITLE
[AG-351] add saddle simulation logic

### PR DESCRIPTION
**Pull request by it4rb**
(used for saddle and iron-stable, oneswap, nerve)

### CODE REVIEW

File: pkg/source/saddle/pool_simulator.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+		reserves[i] = utils.NewBig10(entityPool.Reserves[i])
+		multipliers[i] = utils.NewBig10(staticExtra.PrecisionMultipliers[i])
+	}
+
+	swapFee := utils.NewBig10(extra.SwapFee)
+
+	// only have withdrawFee in saddle v1, default to 0
+	defaultWithdrawFee := utils.NewBig10(extra.DefaultWithdrawFee)
+	if defaultWithdrawFee == nil {
+		defaultWithdrawFee = constant.ZeroBI
```
Comment (piavgh) [inline pkg/source/saddle/pool_simulator.go:56]: looks like there is a mistake here
```suggestion
		defaultWithdrawFee = bignumber.ZeroBI
```

File: pkg/source/saddle/pool_simulator.go [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+				t.Info.SwapFee,
+				t.DefaultWithdrawFee,
+				t.LpSupply,
+				tokenIndexTo,
+				tokenAmountIn.Amount,
+			)
+			if err != nil {
+				return &pool.CalcAmountOutResult{}, err
+			}
+			if amountOut.Cmp(constant.ZeroBI) > 0 {
```
Comment (piavgh) [inline pkg/source/saddle/pool_simulator.go:111]:
```suggestion
			if amountOut.Cmp(bignumber.ZeroBI) > 0 {
```

File: pkg/source/saddle/pool_simulator.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+				t.InitialA,
+				t.DefaultWithdrawFee,
+				t.LpSupply,
+				tokenIndexFrom,
+				tokenAmountIn.Amount,
+			)
+			if err != nil {
+				return &pool.CalcAmountOutResult{}, err
+			}
+			if amountOut.Cmp(constant.ZeroBI) > 0 {
```
Comment (piavgh) [inline pkg/source/saddle/pool_simulator.go:144]:
```suggestion
			if amountOut.Cmp(bignumber.ZeroBI) > 0 {
```

File: pkg/source/saddle/pool_simulator.go (hunk 1 above)
Comment (piavgh) [reply pkg/source/saddle/pool_simulator.go:56]: oh, you alias `bignumber` to `constant`.
It's a bit confusing.

I don't have a strong opinion on this, just make sure to keep it consistent

File: pkg/source/saddle/pool_simulator.go [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+		if t.Info.Tokens[i] == input.Token {
+			t.Info.Reserves[i] = new(big.Int).Add(t.Info.Reserves[i], inputAmount)
+		}
+		if t.Info.Tokens[i] == output.Token {
+			t.Info.Reserves[i] = new(big.Int).Sub(t.Info.Reserves[i], outputAmount)
+		}
+	}
+}
+
+func (t *Pool) GetLpToken() string {
```
Comment (piavgh) [inline pkg/source/saddle/pool_simulator.go:237]: this method is not used anywhere in both `kyberswap-aggregator` and `router-service` so I guess we can remove it

**Review by piavgh — APPROVED** [review]

File: pkg/source/saddle/pool_simulator.go (hunk 1 above)
Comment (it4rb) [reply pkg/source/saddle/pool_simulator.go:56]: yes, sorry for the confusing, I think using alias will reduce the number of change need to be made

**Pull request merged**